flink-sql-runner/src/main/java/com/github/streamshub/flink/SqlRunner.java

```diff
@@ -74,44 +84,101 @@ public static void main(String[] args) throws Exception {
         }
     }
 
-    static List<String> parseStatements(String rawStatements) {
-        var formatted = formatSqlStatements(rawStatements.trim());
+    static List<String> parseStatementArg(String rawStatementArg) {
 
-        var statements = new ArrayList<String>();
-        StringBuilder current = new StringBuilder();
-        Matcher matcher = STATEMENT_SET_PATTERN.matcher(formatted);
+        LOG.debug("Parsing raw statement:\n<begin>{}<end>", rawStatementArg);
 
-        String statementSet = "";
-        String otherStatements = formatted;
+        var statements = new ArrayList<String>();
 
-        if (matcher.find()) {
-            statementSet = matcher.group(1);
-            otherStatements = formatted.replace(statementSet, "").trim();
-        }
+        var cleaned = cleanSqlStatement(rawStatementArg);
+        var formatted = formatSqlStatements(cleaned);
+
+        var statementSetString = new StringBuilder();
+        boolean insideStatementSet = false;
+        // Split the statements on `;` except where that `;` is preceded by a double backspace
+        for (String statement: STATEMENT_DELIMETER_PATTERN.split(formatted)) {
+
+            if (!statement.isBlank()) {
+                // As we split on `;` we need to replace them at the end of the statement.
+                // There is probably a regex incantation we can add to the STATEMENT_DELIMETER_PATTERN to do this, but this will do for now.
+                statement = statement.trim() + STATEMENT_DELIMITER;
+                // Deal with the specific situation where secret strings in WITH clauses contain the `\\;` literal and will end up leaving a `\` before the `;`
+                // Again, there is probably regex foo to achieve this
+                statement = statement.replaceAll(Pattern.quote("\\;"), ";");
+            } else {
+                // If the statement is a blank string then skip to the next one
+                continue;
+            }
 
-        boolean escaped = false;
-        for (char c : otherStatements.toCharArray()) {
-            if (escaped) {
-                current.append(c);
-                escaped = false;
-            } else if (c == '\\') {
-                escaped = true;
-            } else if (c == STATEMENT_DELIMITER.charAt(0)) {
-                current.append(c);
-                statements.add(current.toString().trim());
-                current = new StringBuilder();
+            var statementSetStartMatch = STATEMENT_SET_START_PATTERN.matcher(statement);
+
+            if (statementSetStartMatch.find()) {
+                LOG.debug("Found start of statement set: <begin>{}<end>", statement);
+                insideStatementSet = true;
+                statementSetString.append(statement);
+            } else if (insideStatementSet) {
```

Comment:
Hmm took me a bit of thinking to realise these two blocks can't collapse into one (given `insideStatementSet` is true when `statementSetStartMatch.find()` is true) because the become true on sequential evaluations of the loop. 

I get its logically the same thing (and sonar might whine) but it might be better to finish the `if` block with a `continue`.

Reply:
I added an else block and some more descriptive comments to hopefully make it clear what is going on.